// ngx-drag-drop/esm5/dnd-handle.directive.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Directive, HostBinding, HostListener } from "@angular/core";
import { DndDraggableDirective } from "./dnd-draggable.directive";
import * as ɵngcc0 from '@angular/core';
var DndHandleDirective = /** @class */ (function () {
    function DndHandleDirective(parent) {
        this.draggable = true;
        parent.registerDragHandle(this);
    }
    /**
     * @param {?} event
     * @return {?}
     */
    DndHandleDirective.prototype.onDragEvent = /**
     * @param {?} event
     * @return {?}
     */
    function (event) {
        event._dndUsingHandle = true;
    };
    /** @nocollapse */
    DndHandleDirective.ctorParameters = function () { return [
        { type: DndDraggableDirective }
    ]; };
    DndHandleDirective.propDecorators = {
        draggable: [{ type: HostBinding, args: ["attr.draggable",] }],
        onDragEvent: [{ type: HostListener, args: ["dragstart", ["$event"],] }, { type: HostListener, args: ["dragend", ["$event"],] }]
    };
DndHandleDirective.ɵfac = function DndHandleDirective_Factory(t) { return new (t || DndHandleDirective)(ɵngcc0.ɵɵdirectiveInject(DndDraggableDirective)); };
DndHandleDirective.ɵdir = ɵngcc0.ɵɵdefineDirective({ type: DndHandleDirective, selectors: [["", "dndHandle", ""]], hostVars: 1, hostBindings: function DndHandleDirective_HostBindings(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵlistener("dragstart", function DndHandleDirective_dragstart_HostBindingHandler($event) { return ctx.onDragEvent($event); })("dragend", function DndHandleDirective_dragend_HostBindingHandler($event) { return ctx.onDragEvent($event); });
    } if (rf & 2) {
        ɵngcc0.ɵɵattribute("draggable", ctx.draggable);
    } } });
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(DndHandleDirective, [{
        type: Directive,
        args: [{
                selector: "[dndHandle]"
            }]
    }], function () { return [{ type: DndDraggableDirective }]; }, { draggable: [{
            type: HostBinding,
            args: ["attr.draggable"]
        }], onDragEvent: [{
            type: HostListener,
            args: ["dragstart", ["$event"]]
        }, {
            type: HostListener,
            args: ["dragend", ["$event"]]
        }] }); })();
    return DndHandleDirective;
}());
export { DndHandleDirective };
if (false) {
    /** @type {?} */
    DndHandleDirective.prototype.draggable;
}
